src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.cs

```diff
@@ -9,56 +10,384 @@ namespace Microsoft.Maui.Controls.Platform
+
+		Page? CurrentPage =>
+			_modalPages.Count > 0 ? _modalPages[_modalPages.Count - 1].Page : _window.Page;
+
+		// Shell takes care of firing its own Modal life cycle events
+		// With shell you cam remove / add multiple modals at once
+		bool FireLifeCycleEvents => _window?.Page is not Shell;
+
+		partial void InitializePlatform();
 
 		public ModalNavigationManager(Window window)
 		{
 			_window = window;
+			_window.PropertyChanged += (_, args) =>
+			{
+				if (args.Is(Window.PageProperty))
+					SettingNewPage();
+			};
+
+			InitializePlatform();
+
+			_window.HandlerChanging += OnWindowHandlerChanging;
+			_window.Destroying += (_, _) =>
+			{
+				ClearModalPages(platform: true);
+			};
 		}
 
-		public Task<Page> PopModalAsync()
+		void OnWindowHandlerChanging(object? sender, HandlerChangingEventArgs e)
 		{
-			return PopModalAsync(true);
+			// If the window handler is changing the activity is being recreated
+			// the window activated/resumed event will take care of syncing the platform modals
+			if (e.OldHandler is not null)
+			{
+				ClearModalPages(platform: true);
+			}
 		}
 
+		public Task<Page?> PopModalAsync()
+		{
+			return PopModalAsync(true);
+		}
 
 		public Task PushModalAsync(Page modal)
 		{
 			return PushModalAsync(modal, true);
 		}
 
-		internal void SettingNewPage()
+		bool syncing = false;
+
+		bool IsModalReady
 		{
-			if (_previousPage != null)
+			get
 			{
-				// if _previousNavModel has been set than _navModel has already been reinitialized
-				if (_previousNavModel != null)
+				return
+					_window?.Page?.Handler is not null &&
+					_window.Handler is not null
+					&& IsModalPlatformReady;
+			}
+		}
+
+		void SyncPlatformModalStack([CallerMemberName] string? callerName = null)
+		{
+			var logger = _window.FindMauiContext(true)?.Services?.CreateLogger<ModalNavigationManager>();
+			SyncPlatformModalStackAsync().FireAndForget(logger, callerName);
+		}
+
+		void SyncModalStackWhenPlatformIsReady([CallerMemberName] string? callerName = null)
+		{
+			var logger = _window.FindMauiContext(true)?.Services?.CreateLogger<ModalNavigationManager>();
+			SyncModalStackWhenPlatformIsReadyAsync().FireAndForget(logger, callerName);
+		}
+
+
+		// This code only processes a single sync action per call.
+		// It recursively calls itself until no more sync actions are left to perform.
+		//
+		// A lot can change during the process of pushing/popping a page
+		// i.e. Users might change the root page during an appearing event.
+		// So, instead of just bull dozing through the whole sync we perform one
+		// sync step then recalculate the state of affairs and then perform another
+		// until no more sync operations are left.
+		// Typically it's always a good idea to re-evaluate after any async operation has completed
+		async Task SyncPlatformModalStackAsync()
+		{
+			if (!IsModalReady || syncing)
+				return;
+
+			bool syncAgain = false;
```

Comment:
Wouldn't a counter be better here?

Reply:
What would it count? The main point of this bool is just to indicate that the sync operation produced a change so you should revalidate if you are still in sync with user expectations. I don't really need to count anything. The code just reruns the sync operation recursively until the `platform` matches the `xplat` expectation